activerecord/lib/active_record/migration.rb

```diff
@@ -579,30 +580,41 @@ def self.current_version
     class CheckPending
       def initialize(app, file_watcher: ActiveSupport::FileUpdateChecker)
         @app = app
-        @needs_check = true
-        @mutex = Mutex.new
         @file_watcher = file_watcher
       end
 
       def call(env)
-        @mutex.synchronize do
-          @watcher ||= build_watcher do
-            @needs_check = true
-            ActiveRecord::Migration.check_pending!(connection)
-            @needs_check = false
-          end
-
-          if @needs_check
-            @watcher.execute
-          else
-            @watcher.execute_if_updated
-          end
+        if force_pending_migration_check?
+          watcher.execute
+        else
+          watcher.execute_if_updated
         end
 
         @app.call(env)
       end
 
       private
+        def force_pending_migration_check_flags
+          @force_pending_migration_check_flags ||= Hash.new true
+        end
+
+        def force_pending_migration_check=(value)
+          force_pending_migration_check_flags[WeakRef.new Thread.current] = value
```

Comment:
Why the WeakRef here? Just to avoid having to delete keys from the hash? The size of this cache should not increase with bound given it is direct related to how many threads you use in your webservers.

Reply:
It's true that in Rails' default puma config, the number of threads is static

```ruby
max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
threads min_threads_count, max_threads_count
```

But users can use different configurations which allow puma to kill idle threads and spin up new threads on demand

```ruby
threads 1, 5
```

...or they could be using totally different web servers that similarly spin up and spin down threads as necessary.

Without the `WeakRef`s, that would result in a memory leak since the `Thread` objects would never be GC'd.

Reply:
@rafaelfranca Lemme know if that explanation makes sense. Happy to remove `WeakRef` if it turns out it's not actually beneficial, but I think it might be.